File: src/0328.odd-even-linked-list.328/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function oddEvenList (head: ListNode | null): ListNode | null {
    const headOdd = new ListNode(0);
    let tailOdd = headOdd;
    const headEven = new ListNode(0);
    let tailEven = headEven;
    while (head) {
        const odd = head;
        const even = head.next;
        const next = even ? even.next : null;
        tailOdd.next = odd;
        tailOdd = odd;
        tailOdd.next = null;
        if (even) {
            tailEven.next = even;
            tailEven = even;
            even.next = null;
        }
        head = next;
    }
    tailOdd.next = headEven.next;
    return headOdd.next;
}

File: src/0328.odd-even-linked-list.328/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        ListNode oddDummyHead = new ListNode(0);
        ListNode evenDummyHead = new ListNode(0);
        ListNode oddTail = oddDummyHead;
        ListNode evenTail = evenDummyHead;
        boolean isOdd = true;
        while (head != null){
            ListNode next = head.next;
            head.next = null;
            if(isOdd){
                oddTail.next = head;
                oddTail = head;
            }else{
                evenTail.next = head;
                evenTail = head;
            }
            isOdd = !isOdd;
            head = next;
        }
        oddTail.next = evenDummyHead.next;
        return oddDummyHead.next;
    }
}